Reduce Enum.GetEnumName overheads

stephentoub opened the pull request:
This also helps Enum.ToString(), Enum.IsDefined, etc.  Many enums are composed of sequential values starting at 0 (half of all enums exposed from corelib, for example).  Today when looking up a value, we search the array of values, but if an enum is known to have such sequential values, we can instead just index into the array at the appropriate location.

```C#
private DayOfWeek _dow = DayOfWeek.Saturday;

[Benchmark]
public string DOWSaturday() => Enum.GetName(_dow);
```

|      Method |         Toolchain |      Mean |     Error |    StdDev | Ratio | Code Size |
|------------ |------------------ |----------:|----------:|----------:|------:|----------:|
| DOWSaturday | \main\corerun.exe | 13.271 ns | 0.3248 ns | 0.5427 ns |  1.00 |   1,383 B |
| DOWSaturday |   \pr\corerun.exe |  7.111 ns | 0.2060 ns | 0.4065 ns |  0.54 |     808 B |

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Enum.EnumInfo.cs:
```diff
@@ -17,6 +18,17 @@ public EnumInfo(bool hasFlagsAttribute, ulong[] values, string[] names)
+
+                // Store whether all of the values are sequential starting from zero.
+                ValuesAreSequentialFromZero = true;
+                for (int i = 0; i < values.Length; i++)
```
Isn't there some vector operation that could do this too?

gfoidl replied:
There is, but 
* it is quite a bit amount of code
* the `EnumInfo` gets cached, so not called very often
* Vector128 for `ulong` takes 2 elements, so speedup won't be that big

stephentoub committed:
Reduce Enum.GetEnumName overheads

This also helps Enum.ToString(), Enum.IsDefined, etc.  Many enums are composed of sequential values starting at 0 (half of all enums in corelib, for example, meet this criteria).  Today when looking up a value, we search the array of values, but if an enum is known to have such sequential values, we can instead just index into the array at the appropriate location.

jkotas reviewed (approved):
If you are looking into squeezing more cycles from Enum formatting and parsing, you can try switching on underlying enum CorElementType as the very first thing (using a cached CorElementType for non-generic methods and intrinsic method for the generic methods), and then dispatch to specialized code for each underlying CorElementType.

The benefits of this approach should be:
- Avoid conversions to / from ulong type
- The array of values will be 2x smaller for the most common `int` underlying type
- Eliminate generic code duplication for Enum methods since the switch will completely evaporate and the rest of the code will be all shared between different Enum types

This would mimic what the high-performance Enum packages out there use, except they typically do that using generic code stamping of the whole thing. Sharing the code per CorElementType should solve that problem.

stephentoub committed:
Merge branch 'main' into enumgetname

stephentoub committed:
Address PR feedback

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Enum.cs:
```diff
@@ -116,12 +116,25 @@ private static string ValueToHexString(object value)
+                string[] names = enumInfo.Names;
+                if (ulValue < (ulong)names.Length)
+                {
+                    // TODO https://github.com/dotnet/runtime/issues/76198: Remove use of Unsafe.
```
If memory serves it was maybe 5-10%.  I can undo it, though, and when it's fixed in the JIT this will just get better.

drieseng commented:
@stephentoub Would it possible to also support this for "sequential" enums starting from another value (eg. one), and keep that offset around? I generally prefer to use one for the first value as I don't want a (not nullable) enum field or property to have a (valid) default value if - for some reason - it was not assigned (while it should have been). Perhaps this is not common enough for you to care - and pay the extra cost of keeping an offset around using it where applicable - though, no problem.

stephentoub commented:
> Perhaps this is not common enough for you to care - and pay the extra cost of keeping an offset around using it where applicable - though, no problem.

Yeah, I'm not sure it's worth it. That's not to say we'd never do it, but it would add cost for everyone, and enums with sequential values starting at 1 are a lot less common than with values starting at 0.

danmoseley commented:
Do we have test coverage for both paths?

drieseng commented:
@stephentoub, would you consider a PR where the **ValuesAreSequentialFromZero** bool is replaced with an **ulong?** (or *uint?*) whose value would be would be null if the values are not sequential? Not asking you to guarantee that it would be merged of course.

stephentoub commented:
That's the same as "keep that offset around" that you asked about earlier :)

stephentoub commented:
> Do we have test coverage for both paths?

Yes

drieseng commented:
@stephentoub, yeah indeed. I just meant to say that it's not necessary to have both the offset and the boolean, I wasn't clear enough. If I find a moment, I'll prepare a PR and we'll see how it affects performance (for both cases).

stephentoub commented:
>  just meant to say that it's not necessary to have both the offset and the boolean

A `ulong?` is a `Nullable<ulong>` which is a `ulong` and a `bool` :smile:

drieseng commented:
You can indeed see it that way. As I said, I'll see if I can prepare a PR. If it doesn't meet the bar, that's ok.
